mkdocs/nav.py

```diff
@@ -14,18 +14,33 @@
 log = logging.getLogger(__name__)
 
 
-def filename_to_title(filename):
+def file_to_title(filename):
     """
     Automatically generate a default title, given a filename.
     """
     if utils.is_homepage(filename):
         return 'Home'
 
+    table_of_contents = None
+    try:
+        with open(filename, 'r') as f:
+            lines = f.read()
+            html_content, table_of_contents, meta = utils.convert_markdown(lines, ['toc'])
+            if len(table_of_contents.items) > 0:
```

Comment:
Why would I want to define meta-data bout a file separate from the file? I assumes that the pages list provided for defining the title because it was less expensive than reading the file twice. But if you are going to read the file twice (or at least once while building the nav), then I absolutely expect all meta-data to be defined in the file's meta-data. In my experience other static site generators offer this as an option. The meta-data is checked first, then a fallback to other methods is used. For some reason I don't understand, it seems that most users depend on the fallbacks.